### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by SamMousa**
This is a big cleanup PR that fixes a plethora of smaller issues and rewrites other parts.

| Q             | A
| ------------- | ---
| Breaks BC?    | ✔️ big time

### CODE REVIEW

**Comment by SamMousa** [comment]
Apologies for the lack of specificity but the library was so small that a quick rewrite that breaks BC was more efficient. Hope you have time to review and provide feedback.

Some things this PR does:

- Implement interfaces for message signing en encrypting, something Symfony lacks.
- Implement wrappers for Symfony's signers & encrypters that implement the new interfaces.
- Implement interface `SymfonyMessageWrapperInterface` which contains the `getSymfonyEmail()` function that our `Mailer` needs. This allows us to typehint against the interface not the class.
- Rewrote `Logger` to use the PSR provided trait, allowing us to implement just the log function. Added configurable severity mappings and configurable category.
- Removed `with-` functions in the `Mailer`, Yii components are traditionally mutable and reconfigurable so this pattern is not needed.
- Allow user to set their own `TransportFactory` this allows them to configure things like a custom `HttpClient` but also available transports and which `Dispatcher`, if any, to use.
- Added test coverage, almost all code is now fully covered, except `Message` which sits at ~80%. The proxy wrappers that just contain 2 lines of code like `DkimSignerWrapper` have been excluded from code coverage.

File: src/Mailer.php [hunk 1]
```diff
@@ -72,47 +73,42 @@ public function getSymfonyMailer(): SymfonyMailer
...
         }
 
+        $this->_transport = $transport instanceof TransportInterface ? $transport : $this->createTransport($transport);
+
         $this->symfonyMailer = null;
     }
 
     /**
-     * @return TransportInterface
+     * @deprecated This will become private
```
Comment (samdark) [inline src/Mailer.php:86]: What do you mean by "will become"?

File: src/SMimeEncrypterWrapper.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace yii\symfonymailer;
+
+use Symfony\Component\Mime\Crypto\SMimeEncrypter;
+use Symfony\Component\Mime\Message;
+
+/**
+ * @codeCoverageIgnore This class is a trivial proxy that requires no testing
+ */
+class SMimeEncrypterWrapper implements SymfonyMessageEncrypterInterface
```
Comment (samdark) [inline src/SMimeEncrypterWrapper.php:13]:
```suggestion
class SMimeMessageEncrypter implements SymfonyMessageEncrypterInterface
```

File: src/SMimeSignerWrapper.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace yii\symfonymailer;
+
+use Symfony\Component\Mime\Crypto\SMimeSigner;
+use Symfony\Component\Mime\Message;
+
+/**
+ * @codeCoverageIgnore This class is a trivial proxy that requires no testing
+ */
+class SMimeSignerWrapper implements SymfonyMessageSignerInterface
```
Comment (samdark) [inline src/SMimeSignerWrapper.php:13]:
```suggestion
class SMimeMessageSigner implements SymfonyMessageSignerInterface
```

File: src/SymfonyMessageEncrypterInterface.php [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace yii\symfonymailer;
+
+use Symfony\Component\Mime\Message as Message;
+
+interface SymfonyMessageEncrypterInterface
```
Comment (samdark) [inline src/SymfonyMessageEncrypterInterface.php:9]:
```suggestion
interface MessageEncrypterInterface
```

File: src/SymfonyMessageSignerInterface.php [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace yii\symfonymailer;
+
+use Symfony\Component\Mime\Message;
+
+interface SymfonyMessageSignerInterface
```
Comment (samdark) [inline src/SymfonyMessageSignerInterface.php:9]:
```suggestion
interface MessageSignerInterface
```

File: src/Mailer.php (hunk 1 above)
Comment (SamMousa) [reply src/Mailer.php:86]: Hmm that's a remnant of when I thought I'd do a small non-bc-breaking PR ;-) will fix it.

**Comment by Krakozaber** [comment]
Is the completion of the work planned ?

**Comment by SamMousa** [comment]
Yes definitely. This branch is already in production so I'm pretty sure it works.
Needs a good review though.

**Comment by samdark** [comment]
@SamMousa shouldn't readme be adjusted?

**Comment by SamMousa** [comment]
Hmm, good point, very probably yes. Will try to make some time to finish this!

**Comment by Krakozaber** [comment]
Is everything bad?

**Comment by SamMousa** [comment]
Well it depends on what year we live in. When Yii2 came out it was good, but that was a long time ago. Nowadays the standard "Yii" way is not considered a clean / proper way of writing code. (Not blaming anyone, back then we had php 5.6, no typing, no static analysis etc)
Modern code is different and practices have improved as well.

**Comment by Krakozaber** [comment]
What needs to be completed for the merger?

**Comment by SamMousa** [comment]
The Readme

**Comment by Krakozaber** [comment]
@SamMousa Is there any way I can help with this?

**Comment by SamMousa** [comment]
Definitely, just check the current Readme and make sure it's code samples are still correct!

**Comment by SamMousa** [comment]
@samdark as far as I can tell the readme is good, since the public configuration API is mostly unchanged.

This means this can be merged. In my opinion: release it as a new major version, it may require changes in consuming code if they use advanced stuff like message signing or encryption.

File: src/Logger.php [hunk 6]
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
...
-            case 'warning':
-                $level = \yii\log\Logger::LEVEL_WARNING;
-                break;
-            case 'debug':
-            case 'info':
-                $level = \yii\log\Logger::LEVEL_INFO;
-                break;
-            default:
-                $level = \yii\log\Logger::LEVEL_INFO;
+        if (! is_string($level)) {
```
Comment (samdark) [inline src/Logger.php:60]:
```suggestion
        if (!is_string($level)) {
```

File: src/Logger.php [hunk 7]
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
...
+            throw new \InvalidArgumentException("This logger only supports string levels");
         }
-        Yii::getLogger()->log($message, $level, __METHOD__);
+        if (! isset($this->map[$level])) {
```
Comment (samdark) [inline src/Logger.php:63]:
```suggestion
        if (!isset($this->map[$level])) {
```

File: src/Mailer.php [hunk 8]
```diff
@@ -58,158 +62,100 @@ private function createSymfonyMailer(): SymfonyMailer
...
     public function setTransport($transport): void
     {
-        if (!is_array($transport) && !$transport instanceof TransportInterface) {
-            throw new InvalidConfigException('"' . get_class($this) . '::transport" should be either object or array, "' . gettype($transport) . '" given.');
-        }
-        if ($transport instanceof TransportInterface) {
-            $this->_transport = $transport;
-        } elseif (is_array($transport)) {
-            $this->_transport = $this->createTransport($transport);
+        if (! is_array($transport) && ! $transport instanceof TransportInterface) {
```
Comment (samdark) [inline src/Mailer.php:79]:
```suggestion
        if (!is_array($transport) && !$transport instanceof TransportInterface) {
```

File: src/Mailer.php [hunk 9]
```diff
@@ -58,158 +62,100 @@ private function createSymfonyMailer(): SymfonyMailer
...
-    /**
-     * @return TransportInterface
-     */
-    public function getTransport(): TransportInterface
+    private function getTransport(): TransportInterface
     {
-        if (!is_object($this->_transport)) {
-            $this->_transport = $this->createTransport($this->_transport);
+        /** @psalm-suppress RedundantPropertyInitializationCheck Yii2 configuration flow does not guarantee full initialisation */
+        if (! isset($this->_transport)) {
```
Comment (samdark) [inline src/Mailer.php:91]:
```suggestion
        if (!isset($this->_transport)) {
```

File: src/Logger.php [hunk 10]
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
...
-                $level = \yii\log\Logger::LEVEL_WARNING;
-                break;
-            case 'debug':
-            case 'info':
-                $level = \yii\log\Logger::LEVEL_INFO;
-                break;
-            default:
-                $level = \yii\log\Logger::LEVEL_INFO;
+        if (! is_string($level)) {
+            throw new \InvalidArgumentException("This logger only supports string levels");
```
Comment (samdark) [inline src/Logger.php:61]:
```suggestion
            throw new \InvalidArgumentException("This logger only supports string levels.");
```

File: src/Logger.php [hunk 11]
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
...
         }
-        Yii::getLogger()->log($message, $level, __METHOD__);
+        if (! isset($this->map[$level])) {
+            throw new InvalidArgumentException("Unknown logging level $level");
```
Comment (samdark) [inline src/Logger.php:64]:
```suggestion
            throw new InvalidArgumentException("Unknown logging level $level.");
```

File: src/Mailer.php [hunk 12]
```diff
@@ -58,158 +62,100 @@ private function createSymfonyMailer(): SymfonyMailer
...
-    /**
-     * {@inheritDoc}
-     *
-     * @throws TransportExceptionInterface If sending failed.
-     */
     protected function sendMessage($message): bool
     {
-        if (!($message instanceof Message)) {
-            throw new RuntimeException(sprintf(
+        if (! ($message instanceof MessageWrapperInterface)) {
```
Comment (samdark) [inline src/Mailer.php:136]:
```suggestion
        if (!($message instanceof MessageWrapperInterface)) {
```

File: src/Message.php [hunk 13]
```diff
@@ -153,122 +192,79 @@ public function setSender(string $address): self
...
-    public function embedContent($content, array $options = [])
+    public function embedContent($content, array $options = []): string
     {
-        $file = [];
-        if (!empty($options['fileName'])) {
-            $file['name'] = $options['fileName'];
-        } else {
-            $file['name'] = null;
+        if (empty($options['fileName'])) {
+            throw new InvalidConfigException('A valid file name must be passed when embedding content');
```
Comment (samdark) [inline src/Message.php:259]:
```suggestion
            throw new InvalidConfigException('A valid file name must be passed when embedding content.');
```

File: src/Message.php [hunk 14]
```diff
@@ -352,14 +353,10 @@ private function convertAddressesToStrings(array $addresses)
...
-        if (is_string($strings)) {
-            return [new Address($strings)];
-        }
-
         $addresses = [];
 
-        foreach ($strings as $address => $name) {
-            if (!is_string($address)) {
+        foreach ((array) $strings as $address => $name) {
+            if (! is_string($address)) {
```
Comment (samdark) [inline src/Message.php:359]:
```suggestion
            if (!is_string($address)) {
```

File: tests/MessageTest.php [hunk 15]
```diff
@@ -14,24 +16,24 @@
...
         $this->mockApplication([
             'components' => [
-                'mailer' => $this->createTestEmailComponent()
-            ]
+                'mailer' => $this->createTestEmailComponent(),
+            ],
         ]);
         $filePath = $this->getTestFilePath();
-        if (!file_exists($filePath)) {
+        if (! file_exists($filePath)) {
```
Comment (samdark) [inline tests/MessageTest.php:36]:
```suggestion
        if (!file_exists($filePath)) {
```

File: tests/MessageTest.php [hunk 16]
```diff
@@ -78,9 +80,9 @@ protected function createTestMessage()
      * @param  string $text     text to be applied on image.
      * @return string image file full name.
      */
-    protected function createImageFile($fileName = 'test.jpg', $text = 'Test Image')
+    private function createImageFile($fileName = 'test.jpg', $text = 'Test Image'): string
     {
-        if (!function_exists('imagecreatetruecolor')) {
+        if (! function_exists('imagecreatetruecolor')) {
```
Comment (samdark) [inline tests/MessageTest.php:85]:
```suggestion
        if (!function_exists('imagecreatetruecolor')) {
```

File: tests/TestCase.php [hunk 17]
```diff
@@ -34,7 +34,7 @@ protected function mockApplication($config = [], $appClass = '\yii\console\Appli
     protected function getVendorPath()
     {
         $vendor = dirname(dirname(__DIR__)) . '/vendor';
-        if (!is_dir($vendor)) {
+        if (! is_dir($vendor)) {
```
Comment (samdark) [inline tests/TestCase.php:37]:
```suggestion
        if (!is_dir($vendor)) {
```

File: tests/compatibility.php [hunk 18]
```diff
@@ -4,14 +4,17 @@
  */
 
 namespace PHPUnit\Framework\Constraint {
-    if (!class_exists('PHPUnit\Framework\Constraint\Constraint') && class_exists('PHPUnit_Framework_Constraint')) {
-        abstract class Constraint extends \PHPUnit_Framework_Constraint {}
+    if (! class_exists('PHPUnit\Framework\Constraint\Constraint') && class_exists('PHPUnit_Framework_Constraint')) {
```
Comment (samdark) [inline tests/compatibility.php:7]:
```suggestion
    if (!class_exists('PHPUnit\Framework\Constraint\Constraint') && class_exists('PHPUnit_Framework_Constraint')) {
```

File: tests/compatibility.php [hunk 19]
```diff
@@ -4,14 +4,17 @@
...
+        abstract class Constraint extends \PHPUnit_Framework_Constraint
+        {
+        }
     }
 }
 
 namespace PHPUnit\Framework {
-    if (!class_exists('PHPUnit\Framework\TestCase') && class_exists('PHPUnit_Framework_TestCase')) {
-        abstract class TestCase extends \PHPUnit_Framework_TestCase {
+    if (! class_exists('PHPUnit\Framework\TestCase') && class_exists('PHPUnit_Framework_TestCase')) {
```
Comment (samdark) [inline tests/compatibility.php:15]:
```suggestion
    if (!class_exists('PHPUnit\Framework\TestCase') && class_exists('PHPUnit_Framework_TestCase')) {
```

**Review by samdark — APPROVED** [review]
👍 Need a line for CHANGELOG.

**Comment by SamMousa** [comment]
@samdark your code style changes should be enforced by easy coding standard. (Check `ecs.php`) 

I can use a different standard / config, but the idea is to automate the check so that code style feedback by humans isn't needed and any PR will fail if it doesn't abide by the rules.

**Comment by samdark** [comment]
I know. Doing that in Yii3 already via StyleCI.

**Comment by nd4c** [comment]
I've just quickly reviewed some of the changes.  I'm concerned that getSymfonyMailer() method has been made private.  This makes it impossible to extend the Mailer class with a new sendMessage() method.  Please keep it at least protected.

In fact it would be far nicer if $encryptor, $signer and $dkimSignerOptions were also protected and not private.

**Comment by nd4c** [comment]
Please also not that in the sendMessage method, it needs to re-throw the Exception since the calling code will need that to extract the errors.  No need to return false;

**Comment by nd4c** [comment]
There are a lot of changes in this.  Is it possible to include a public socket timeout property?

https://github.com/yiisoft/yii2-symfonymailer/issues/40

**Comment by SamMousa** [comment]
> In fact it would be far nicer if $encryptor, $signer and $dkimSignerOptions were also protected and not private.

They are public as far as I can tell. So you can configure them however you like.

**Comment by SamMousa** [comment]
> There are a lot of changes in this. Is it possible to include a public socket timeout property?

This is not really related to these changes. But the way to go about it (and this is purely the Symfony part of the library) is as you suspected.
You'd use a wrapper class (composition instead of inheritance) since Symfony even goes so far as to make their classes final so you cannot extend them.

```php
class SmtpTransportWithTimeoutFactory implements TransportFactoryInterface {
    public function __construct(private TransportFactoryInterface $factory, private float $timeout)
    {
    }

    public function create(Dsn $dsn): TransportInterface
    {
        $result = $this->factory->create($dsn);
        if ($result instanceof SmtpTransport) {
            $result->getStream()->setTimeout($this->timeout);
        }
        return $result;
    }

    public function supports(Dsn $dsn): bool {
        return $this->factory->supports($dsn);
    }
}

```

**Comment by SamMousa** [comment]
@samdark I think it's really done now!

**Comment by nd4c** [comment]
I really hate to bother you guys. I am really looking forward to these changes

When I call:
`$mailer = Yii::$app->mailer;`
I keep getting a error:

Error 
Typed property ....\Mailer::$transportFactory must not be accessed before initialization

I was trying to set it up so I could start building and wiring up a Transport.  The config is very normal and complies with the ReadMe sample with the exception that the 'dsn' key is not used as this will override the other settings.

    private function createTransport(array $config = []): TransportInterface
    {
        if (array_key_exists('dsn', $config)) {
            $transport = $this->transportFactory->fromString($config['dsn']);
        } elseif (array_key_exists('scheme', $config) && array_key_exists('host', $config)) {
            $dsn = new Dsn(
                $config['scheme'],
                $config['host'],
                $config['username'] ?? '',
                $config['password'] ?? '',
                $config['port'] ?? null,
                $config['options'] ?? [],
            );
            $transport = $this->transportFactory->fromDsnObject($dsn);
        } else {
            throw new InvalidConfigException('Transport configuration array must contain either "dsn", or "scheme" and "host" keys.');
        }
        return $transport;
    }

I have tried different ways of "installing" the new version - automagical loading can get confusing - but with the config containing a transport configuration array it appears that the BaseObject constructor is calling Yii::configure($this, $config) before calling $this->init();

Yii::configure will end up calling setTransport with the configuration array.  And since it's an array setTransport calls createTransport which tries to access $this->transportFactory and that causes an exception:

Typed property ....\Mailer::$transportFactory must not be accessed before initialization

It very possible that I've wired up the Mailer wrong - if that's the case I'm very sorry for wasting your time.

**Comment by nd4c** [comment]
This is a sequential view of what's happening, starting with BaseObject::_construct().

yii\base\BaseObject calls BaseYii::configure() before calling $this->init()

    public function __construct($config = [])
    {
        if (!empty($config)) {
            Yii::configure($this, $config);  [THE CALL]
        }
        $this->init();
    }
 
BaseYii calls yii\base\Component::_set()

    public static function configure($object, $properties)
    {
        foreach ($properties as $name => $value) {
            $object->$name = $value; 
 
Component calls Mailer::setTransport because we have a 'transport' => array configuration.

    public function __set($name, $value)
    {
        $setter = 'set' . $name;
        if (method_exists($this, $setter)) {
            // set property
            $this->$setter($value);
 
Mailer calls its own createTransport because $transport is a configuration array

    public function setTransport($transport): void
    {
        $this->_transport = $transport instanceof TransportInterface ? $transport : $this->createTransport($transport);
 
And createTransport tries to access $this->transportFactory before it's initialized by init()

        } elseif (array_key_exists('scheme', $config) && array_key_exists('host', $config)) {
            $dsn = new Dsn(
                $config['scheme'],
                $config['host'],
                $config['username'] ?? '',
                $config['password'] ?? '',
                $config['port'] ?? null,
                $config['options'] ?? [],
            );
            $transport = $this->transportFactory->fromDsnObject($dsn);

**Comment by SamMousa** [comment]
You must be in a different timezone 😅 I'll have a look. Note that the tests should show an example of how I wired it up. That is not to say this isn't a bug.

**Comment by SamMousa** [comment]
@nd4c give it another go; I've fixed it and added a test case for it.
Funny thing is that this type of bug is normally caught by static analysis, but we had to suppress those initialization checks because of the way object creation works in the framework :-/

**Comment by SamMousa** [comment]
Hmm, should we even support these custom options?

Instead we could support just DSN and reference this doc: https://symfony.com/doc/current/mailer.html#transport-setup

**Comment by nd4c** [comment]
I would recommend allowing them because if someone creates a new Transport that has a lot of 'options' it's much easier to see what they are rather than appending them to the 'dsn' string.

**Comment by nd4c** [comment]
Also, I would recommend using 'protected' instead of 'private'.  This allows the classes to be extended.

Quote from the "About Yii" section of the "Guide":
`Yii is extremely extensible. You can customize or replace nearly every piece of the core's code. You can also take advantage of Yii's solid extension architecture to use or develop redistributable extensions.`

And since this is an official Yii extension rather than an independent library it makes sense to me to follow that "extensible" aspect.

**Comment by SamMousa** [comment]
>  This allows the classes to be extended.

You keep coming back to this, but I believe it is a fundamental misunderstanding of how to write extensible code. Code being extensible does not have to mean that classes are extendable.
Anything you have mentioned wanting to do by using custom configuration or extending classes you can do better by using composition.

> I would recommend allowing them because if someone creates a new Transport that has a lot of 'options' it's much easier to see what they are rather than appending them to the 'dsn' string.

The library we use uses a DSN though, and the DSN object inherently loses features like `failover` or `roundrobin` configuration. Which means that we sacrifice features from the parent library just to introduce our own configuration syntax which we then need to document and help users with. 

Thinking it through a bit more I'm pretty sure that even leaving in the `transport` configuration option is not the best design. 
Ideally you configure the factory and have it support your custom transport. The custom transport is than configured using a DSN like all other transports already supported by the library.

This is way cleaner:
```php
[
    'class' => Mailer::class,
    'dsn' => "sendgrid+api://{$env->get('SENDGRID')}@default"
]
```
And if you need a custom factory implement that:

```php
[
    'class' => Mailer::class,
    'transportFactory' => new Transport(new MyTransportFactory()),
    'dsn' => "my-custom-protocol://{$env->get('SENDGRID')}@default"
]
```

**Comment by nd4c** [comment]
Composition it is then.  Of course, you realize that the Mailer class extends BaseMailer...

I don't see that strictly using a 'dsn' element is necessarily cleaner.  When using certain Transports, yes, it's cleaner.  But not for all Transports.

	'transport' => [
		'scheme' => 'smtps',
		'host' => 'mail.longwinded.com',
		'port' => 2525,
		'username' => 'customersupport@mailingstation.anotherverylongurl.com',
		'password' => 'We prefertousea Verylongandunforgettablepassphrase insteadof a password',
		'options' => [
			'verify_peer' => 0,
			'local_domain' => 'example.org',
			'restart_threshold' =< 10,
			'restart_threshold_sleep' => 1,
			'ping_threshold' => 200,
		],
	],

That would be a very long 'dsn' string.  And, again, the DSN object just parses the string to get back to what's clearly displayed in the array.

**Comment by nd4c** [comment]
I guess if I wanted to use `failover` or `roundrobin` with such complicated transport settings I would probably need a Mailer Factory that could automatically compose the 'dsn' string from a much more readable config array.  So it might make sense to just use the 'dsn' string.  I just keep feeling that we'd be missing a nice feature.

**Comment by SamMousa** [comment]
> Composition it is then. Of course, you realize that the Mailer class extends BaseMailer...

I do, there's no easy way around it. Also realize that the framework is +- 10 years old, it's core code is not a best practice in 2022.

**Comment by Krakozaber** [comment]
[samdark](https://github.com/samdark) added this to the [3.0.0](https://github.com/yiisoft/yii2-symfonymailer/milestone/1) milestone [9 days ago](https://github.com/yiisoft/yii2-symfonymailer/pull/27#event-7318282013)

does that mean never?

**Comment by samdark** [comment]
No. That means it will be a major release of this exact extension.

**Comment by machour** [comment]
@SamMousa thank you for reworking this!

I'm wondering if we could also include [`MailerEvents`](https://symfony.com/doc/current/mailer.html#mailer-events) as Yii events? (definitely needed to inline CSS or stuff like this prior to sending the message)

**Comment by SamMousa** [comment]
> @SamMousa thank you for reworking this!
> 
> I'm wondering if we could also include [`MailerEvents`](https://symfony.com/doc/current/mailer.html#mailer-events) as Yii events? (definitely needed to inline CSS or stuff like this prior to sending the message)

Not in this pr. Let's do that after maybe

**Comment by tgerakitis** [comment]
just curious, what is the ETA on this PR? 
or is it possible to require this branch using composer?

thanks for your hard work! 😊

**Comment by SamMousa** [comment]
> or is it possible to require this branch using composer?

Yes. Google is your friend.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
@SamMousa sorry it took so long. Would you please help closing issues this PR solves? Thanks.

**Comment by Krakozaber** [comment]
@samdark need a tag

**Comment by samdark** [comment]
@SamMousa, @Krakozaber is readme alright?

**Comment by nd4c** [comment]
Also, I believe the 'scheme' settings is better to be left out.  

'scheme' => 'smtps' is kind of misleading.  It doesn't exactly force ssl, but rather causes the code to choose a different transport.

I forget the details, but it's the 'port' => '465' that cause the code to use ssl.  And the default transport works better than the 'smtps' transport.

So the better config would be:

```
'transport' => [
                'host' => '',
                'username' => '',
                'password' => '',
                'port' => 465,
            ],

```

**Comment by nd4c** [comment]
I've never dones a PR before.  Any tips?

**Comment by SamMousa** [comment]
Personally I believe you should be using DSN in all cases. The underlying library uses DSN and by not using it all you do is lose functionality.

> I've never dones a PR before. Any tips?

For these simple things you can just do them inside Github and it will do most of the meta-work for you. 
Go here: https://github.com/yiisoft/yii2-symfonymailer/blob/master/README.md and click the pencil top right.

Good luck!

**Comment by nd4c** [comment]
Actually the underlying library uses the inidividual fields.  It will convert any DSN string to individual attributes on it's DSN object.  And not all transports are easy to for a developer to "read" as a DSN string.  (And for that matter, the DSN string is not technically a DSN, but rather a list of configuration settings...)

I will work on the README.md file.

**Pull request merged**